fix: Input Widget V3 bug fixes

jsartisan opened the pull request:
Fixes #27624

marks0351 commented on app/client/src/pages/Editor/Canvas.tsx:
```diff
@@ -86,7 +87,7 @@ const Canvas = (props: CanvasProps) => {
           data-testid="t--canvas-artboard"
           id="art-board"
-          ref={focusRef}
+          ref={isWDSEnabled ? undefined : focusRef}
```
is there a usecase for this ref?

jsartisan commented:
/build-deploy-preview env=release

ashit-rath commented on app/client/src/widgets/wds/WDSButtonWidget/widget/index.tsx:
```diff
@@ -27,7 +27,7 @@ class ButtonWidget extends BaseWidget<ButtonWidgetProps, ButtonWidgetState> {
     };
   }
 
-  static type = "BUTTON_WIDGET_V2";
```
Wouldn't existing applications using `BUTTON_WIDGET_V2` fail because of modifying this?

ashit-rath commented on app/client/src/widgets/wds/WDSInputWidget/widget/index.tsx:
```diff
@@ -20,16 +20,16 @@ import type { WidgetState } from "widgets/BaseWidget";
 import { EventType } from "constants/AppsmithActionConstants/ActionConstants";
 import { ResponsiveBehavior } from "layoutSystems/common/utils/constants";
 
-class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
```
Is this InputWidgetV3 currently available in production for users to use? or is V2 current available and this is behind feature flag?

jsartisan commented:
/build-deploy-preview env=release

KelvinOm commented on app/client/src/widgets/wds/WDSBaseInputWidget/widget/index.tsx:
```diff
@@ -119,8 +119,8 @@ class WDSBaseInputWidget<
-      ((isMultiLine && (e.metaKey || e.ctrlKey)) ||
-        (!isMultiLine && isEnterKey && isValid))
+      isEnterKey &&
+      ((isMultiLine && (e.metaKey || e.ctrlKey)) || (!isMultiLine && isValid))
```
It's not clear from the code what's going on here. Also could you add a unit test for this?

shadabbuchh commented:
Additional bugs after the latest fixes:

1. Tooltip does not appear upon hover, instead we have to click the icon. Also in order to close the tooltip we have to click the icon again. This is not a good UX as there are additional clicks. Additionally on clicking the icon we sometimes lose focus & the tooltip disappears making it hard for the user to read it.

https://github.com/appsmithorg/appsmith/assets/39921438/d3ae5ed3-cdfd-4eff-8db6-61f17fd6b4f7

2. Max. character issue:
- Enter any text inside the input widget.
- Now set a value for `Max characters` such that it is less that the entered number of characters.
- Now delete the set limit & try to enter more characters inside the widget.

User is not able to enter characters further. Also when the max character limit is exceeded, we do not get the error message in the edit mode.
https://www.loom.com/share/5afeeff5da484458a764f5d6903c209e

3. We do not see the appropriate cursor when the widget is disabled.
4. Upon entering an invalid input, the user has to hover for a long time before they get to see the error message.
5. No error message shown when text is entered instead of number.
https://www.loom.com/share/c2b7d7d42511442e9cc359cc720a9bb2

6. Ellipsis in the long input name vanishes when the user clicks/hovers inside the input area. Also the ellipsis does not appear in deploy mode.

https://github.com/appsmithorg/appsmith/assets/39921438/f486660d-6e5e-46af-8839-5e41c5f2363b

7. Required message behaves weirdly.
https://www.loom.com/share/82503ab48af34914b0af28d50e63957b

jsartisan commented:
/build-deploy-preview env=release

jsartisan commented:
/build-deploy-preview env=release

jsartisan commented:
/build-deploy-preview env=release

jsartisan commented:
/build-deploy-preview env=release skip-tests=true

dhruvikn commented:
/build-deploy-preview skip-tests=true

jsartisan commented:
@shadabbuchh 
1. This is happening because the widget is focused. The focus is going on codemirror input. So the behavious is expected.
2. Fixed
3. Fixed
4. This is something the browser is doing. Not under our control :/
5. Fixed
6. Fixed
7. Fixed

jsartisan commented:
/build-deploy-preview skip-tests=true

shadabbuchh commented:
All reported issues have been fixed.
There are some minor bugs/issues that we can look at a later point in time:

1. When the `Required` prop is enabled, we see a red outline upon entering any text for a fraction of second.
2. Add punctuations for error messages.

Created a separate bug for the above in order to track it.
https://github.com/appsmithorg/appsmith/issues/28186

jsartisan commented:
/ok-to-test
